src/SubtitleOctopus.cpp

```diff
@@ -78,13 +80,114 @@ const float MAX_UINT8_CAST = 255.9 / 255;
 #define CLAMP_UINT8(value) ((value > MIN_UINT8_CAST) ? ((value < MAX_UINT8_CAST) ? (int)(value * 255) : 255) : 0)
 
 typedef struct {
-public:
     int changed;
     double blend_time;
     int dest_x, dest_y, dest_width, dest_height;
     unsigned char* image;
 } RenderBlendResult;
 
+typedef struct {
+    double eventFinish, emptyFinish;
+    int is_animated;
+} EventStopTimesResult;
+
+static int _is_move_tag_animated(char *begin, char *end) {
+    int params[6];
+    int count = 0, value = 0, num_digits = 0;
+    for (; begin < end; begin++) {
+        switch (*begin) {
+            case ' ': // fallthrough
+            case '\t':
+                break;
+            case ',':
+                params[count] = value;
+                count++;
+                value = 0;
+                num_digits = 0;
+                break;
+            default: {
+                    int digit = *begin - '0';
+                    if (digit < 0 || digit > 9) return 0; // invalid move
+                    value = value * 10 + digit;
+                    num_digits++;
+                    break;
+                }
+        }
+    }
+    if (num_digits > 0) {
+        params[count] = value;
+        count++;
+    }
+    if (count < 4) return 0; // invalid move
+
+    // move is animated if (x1,y1) != (x2,y2)
+    return params[0] != params[2] || params[1] != params[3];
+}
+
+static int _is_animated_tag(char *begin, char *end) {
+    // strip whitespaces around the tag
+    while (begin < end && (*begin == ' ' || *begin == '\t')) begin++;
+    while (end > begin && (end[-1] == ' ' || end[-1] == '\t')) end--;
+
+    int length = end - begin;
```

Comment:
I think `begin..end` range will never be zero-terminated.
```suggestion
    int length = end - begin + 1;
```